ffi/rust/src/lib.rs

```diff
@@ -0,0 +1,95 @@
+use std::ffi::{CStr, CString, NulError};
+use std::str::Utf8Error;
+
+use thiserror::Error;
+
+#[allow(non_camel_case_types, non_upper_case_globals, improper_ctypes)]
+pub mod ffi_wrapper;
+
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub host_address: String,
+    pub resolver_address: String,
+    pub user_did: String,
+    pub agent_did: String,
+    pub agent_key_name: String,
+    pub agent_name: String,
+    pub agent_secret: String,
+    pub token_duration: i64,
+}
+
+#[derive(Error, Debug)]
+pub enum IdentityLibError {
+    #[error(transparent)]
+    Utf8Error(#[from] Utf8Error),
+    #[error(transparent)]
+    NulError(#[from] NulError),
+    #[error("FFI library error message: `{0}`.")]
+    Message(String),
+}
+
+impl Drop for ffi_wrapper::CreateAgentAuthToken_return {
+    fn drop(&mut self) {
+        unsafe {
+            ffi_wrapper::FreeUpCString(self.r0);
+            ffi_wrapper::FreeUpCString(self.r1);
+        }
+    }
+}
+
+impl Drop for ffi_wrapper::CreateTwinWithControlDelegation_return {
+    fn drop(&mut self) {
+        unsafe {
+            ffi_wrapper::FreeUpCString(self.r0);
+            ffi_wrapper::FreeUpCString(self.r1);
+        }
+    }
+}
+
+pub fn create_agent_auth_token(options: &Config) -> Result<String, IdentityLibError> {
+    let result = unsafe {
+        ffi_wrapper::CreateAgentAuthToken(
+            go_string(&CString::new(options.agent_did.as_str())?),
+            go_string(&CString::new(options.agent_key_name.as_str())?),
+            go_string(&CString::new(options.agent_name.as_str())?),
+            go_string(&CString::new(options.user_did.as_str())?),
+            go_string(&CString::new(options.agent_secret.as_str())?),
+            options.token_duration,
+        )
+    };
+    let token = unsafe { CStr::from_ptr(result.r0) }.to_str()?;
+    let error = unsafe { CStr::from_ptr(result.r1) }.to_str()?;
+
+    match error.is_empty() {
+        true => Ok(token.to_string()),
+        false => Err(IdentityLibError::Message(error.to_string())),
+    }
+}
+
+pub fn create_twin_with_control_delegation(options: &Config, twin_key_name: &str) -> Result<String, IdentityLibError> {
```

Comment:
Nice to have: It would be good to have some doc comments. You can see the generate docs with `cargo doc --open`.